File: microservices/customer.py
from flask import Flask, request, jsonify
from os import environ
from flask_cors import CORS
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = environ.get('dbURL')
# app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+mysqlconnector://is213@host.docker.internal:3306/property_management'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
app.config['SQLALCHEMY_ENGINE_OPTIONS'] = {'pool_recycle': 299}

db = SQLAlchemy(app)

CORS(app)

class Customer(db.Model):
    __tablename__ = 'customer'

    customer_id = db.Column(db.Integer, db.Sequence('seq_reg_id', start=1, increment=1), primary_key=True)
    name = db.Column(db.String(32), nullable=False)
    phone = db.Column(db.String(8), nullable=False)
    email = db.Column(db.String(20), nullable=False)

    def __init__(self, name, phone, email):
        # self.customer_id = customer_id
        self.name = name
        self.phone = phone
        self.email = email

    def json(self):
        return {"customer_id": self.customer_id, "name": self.name, "phone": self.phone, "email": self.email}



# GET ONE CUSTOMER
@app.route("/customer/<string:customer_id>")
def find_customer(customer_id):
    customer = Customer.query.filter_by(customer_id=customer_id).first()
    if customer:
        return jsonify(
            {
                "code": 200,
                "data": customer.json()
            }
        )
    return jsonify(
        {
            "code": 404,
            "message": "customer not found."
        }
    ), 404

# ADDING A CUSTOMER
@app.route("/customer", methods=['POST'])
def create_customer():
    data = request.get_json()
    customer = Customer(**data)

    try:
        db.session.add(customer)
        db.session.commit()
    except:
        return jsonify(
            {
                "code": 500,
                "message": "An error occurred creating the customer."
            }
        ), 500

    return jsonify(
        {
            "code": 201,
            "data": customer.json()
        }
    ), 201


@app.route("/customer/<string:customer_id>", methods=['PUT'])
def update_customer(customer_id):
    customer = Customer.query.filter_by(customer_id=customer_id).first()
    if customer:
        data = request.get_json()
        if data['name']:
            customer.name = data['name']
        if data['phone']:
            customer.phone = data['phone']
        if data['email']:
            customer.email = data['email'] 
        db.session.commit()
        return jsonify(
            {
                "code": 200,
                "data": customer.json()
            }
        )
    return jsonify(
        {
            "code": 404,
            "data": {
                "customer_id": customer_id
            },
            "message": "customer not found."
        }
    ), 404


@app.route("/customer/<string:customer_id>", methods=['DELETE'])
def delete_customer(customer_id):
    customer = Customer.query.filter_by(customer_id=customer_id).first()
    if customer:
        db.session.delete(customer)
        db.session.commit()
        return jsonify(
            {
                "code": 200,
                "data": {
                    "customer_id": customer_id
                }
            }
        )
    return jsonify(
        {
            "code": 404,
            "data": {
                "customer_id": customer_id
            },
            "message": "customer not found."
        }
    ), 404


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5700, debug=True)

File: microservices/customer.js
const express = require('express');
const bodyParser = require('body-parser');
const app = express();

// Define your routes and database connections here
app.use(bodyParser.json()); // to parse JSON bodies
app.listen(5700, function() {
    console.log('Server listening on port 5700');
});

var mysql = require('mysql2');
const con = mysql.createConnection(process.env.dbURL);

// var con = mysql.createConnection({
//     host: "host.docker.internal",
//     user: "is213",
//     password: "",
//     database: "property_management"
// });

class Customer {
    constructor(customer_id, name, phone, email) {
        this.customer_id = customer_id;
        this.name = name;
        this.phone = phone;
        this.email = email;
    }

    json() {
        return {
            "customer_id": this.customer_id,
            "name": this.name,
            "phone": this.phone,
            "email": this.email
        }
    }
}


//GET CUSTOMER
app.get('/customer/:customer_id', function(req, res) {
    const customer_id = req.params.customer_id;
    con.query("SELECT * FROM customer WHERE customer_id = ?", [customer_id], function (err, result) {
        if (err) throw err;

        if (result.length > 0) {
            const customer = new Customer(result[0].customer_id, result[0].name, result[0].phone, result[0].email);
            res.json({
                "code": 200,
                "data": customer.json()
            });
        } else {
            res.status(404).json({
                "code": 404,
                "message": "customer not found."
            });
        }
    });
});

//ADD AN CUSTOMER
app.post('/customer', function(req, res) {
    console.log(req.body); // log the request body
    const { customer_id, name, phone, email } = req.body;
    const customer = new Customer(customer_id, name, phone, email);

    con.query("INSERT INTO customer SET ?", customer, function (err, result) {
        if (err) {
            console.error(err);
            res.status(500).json({ code: 500, message: "An error occurred creating the customer." });
            return;
        }
        console.log(result);
        res.status(201).json({ code: 201, data: customer.json() });
    });
});

// UPDATE AN CUSTOMER
app.put('/customer/:customer_id', function(req, res) {
    const customer_id = req.params.customer_id;
    const { name, phone, email } = req.body;
    con.query("UPDATE customer SET name = ?, phone = ?, email = ? WHERE customer_id = ?", [name, phone, email, customer_id], function (err, result) {
        if (err) {
            console.error(err);
            console.error(err.message);
            res.status(404).json({
                "code": 404,
                "data": {
                    "customer_id": customer_id
                },
                "message": "customer not found."
            });
            return;
        }
        if (result.affectedRows === 0) {
            res.status(404).json({
                "code": 404,
                "data": {
                    "customer_id": customer_id
                },
                "message": "customer not found."
            });
            return;
        }
        con.query("SELECT * FROM customer WHERE customer_id = ?", [customer_id], function (err, result) {
            if (err) throw err;

            const customer = new Customer(result[0].customer_id, result[0].name, result[0].phone, result[0].email);
            res.json({
                "code": 200,
                "data": customer.json()
            });
        });
    });
});

// DELETE AN CUSTOMER
app.delete('/customer/:customer_id', function(req, res) {
    const customer_id = req.params.customer_id;
    con.query("DELETE FROM customer WHERE customer_id = ?", [customer_id], function (err, result) {
        if (err) {
            console.error(err);
            res.status(500).json({ code: 500, message: "An error occurred deleting the customer." });
            return;
        }
        if (result.affectedRows === 0) {
            res.status(404).json({
                "code": 404,
                "data": {
                    "customer_id": customer_id
                },
                "message": "customer not found."
            });
            return;
        }
        res.json({
            "code": 200,
            "data": {
                "customer_id": customer_id
            }
        });
    });
});


//GET CUSTOMER'S ID FROM EMAIL
app.get('/customer/get_id_by_email/:email', function(req, res) {
    const email = req.params.email;
    con.query("SELECT * FROM customer WHERE email = ?", [email], function (err, result) {
        if (err) throw err;

        if (result.length > 0) {
            const customer = new Customer(result[0].customer_id, result[0].name, result[0].phone, result[0].email);
            res.json({
                "code": 200,
                "data": customer.json()
            });
        } else {
            res.status(404).json({
                "code": 404,
                "message": "customer not found."
            });
        }
    });
});


